tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlockAwareSegmentInputStreamImpl.java

```diff
@@ -161,6 +212,45 @@ private List<ByteBuf> readNextEntriesFromLedger(long start, long maxNumberEntrie
         }
     }
 
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        if (b == null) {
+            throw new NullPointerException("The given bytes are null");
+        } else if (off < 0 || len < 0 || len > b.length - off) {
+            throw new IndexOutOfBoundsException("off=" + off + ", len=" + len + ", b.length=" + b.length);
+        } else if (len == 0) {
+            return 0;
+        }
+
+        int offset = off;
+        int readLen = len;
+        int readBytes = 0;
+        // reading header
+        if (dataBlockHeaderStream.available() > 0) {
+            int read = dataBlockHeaderStream.read(b, off, len);
+            offset += read;
+            readLen -= read;
+            readBytes += read;
+            bytesReadOffset += read;
+        }
+        if (readLen == 0) {
+            return readBytes;
+        }
+
+        // reading ledger entries
+        if (bytesReadOffset < blockSize) {
```

Comment:
Here maybe some problem. If already read data to `b`. But `bytesReadOffset < blockSize), it will return -1.

Reply:
My fault, it won't be happen, ignore.

Reply:
Maybe happen, when the block size < 128, the problem will be happen.

Reply:
And shall we need this logicment, if blockSize < headerSize at initial. When read header stream, the bytesReadOffset will be greater than blockSize. It won't to read entry data.

Reply:
The Header max size is 128. We have a block size limitation for that in the configuration `Max block size in bytes. (64MB by default, 5MB minimum) It looks that is an unnecessary check I think.